std/traits.d

```diff
@@ -7025,6 +7026,41 @@ template Unqual(T)
     static assert(is(Unqual!ImmIntArr == immutable(int)[]));
 }
 
+/**
+Construct a type with a shared tail, and if possible with an unshared head.
+
+If $(D T) does not have indirections, or $(D T) already has an unshared head and
+a shared tail, $(D TailShared) will alias itself away.
+
+---
+// For any type T
+static assert(is(Tailshared!T == TailShared(shared T)));
+---
+*/
+template TailShared(T)
+{
+    // TailShared is already defined privately in core.atomic, instead of
+    // redefining it here, let's just steal it...
+
+    import core.atomic : atomicLoad, MemoryOrder;
+    alias TailShared = ReturnType!(atomicLoad!(MemoryOrder.seq, T));
+}
+
+// NOTE: Full unittest is in core.atomic, This one is mainly for docs.
+///
+@safe unittest
+{
+    static assert(is(TailShared!(shared int) == int));
+    static assert(is(TailShared!(shared int*) == shared(int)*));
+
+    static class C { int i; }
+    static assert(is(TailShared!C == shared C));
+
+    static struct S3 { int* p; int i; }
+    static assert(!is(TailShared!S3 : S3));
+    static assert(is(TailShared!S3 : shared S3));
```

Comment:
Is there a reason you are using the `:` operator instead of `==` inside the is expression? Is there a difference between how `TailShared` works with `class`es and `struct`s?

Reply:
There is indeed a difference between `class` and `struct` types here. A `class` is a reference type, and a `struct` is a value type. So, given `class C` and `struct S`, `shared(C)` means that both the instance itself and its reference are `shared`, while `shared(S)` means the struct instance is `shared`, and it may be referenced either through a `shared` pointer (`shared(S*)`) or an unshared pointer (`shared(S)*`).

So the `==` in the `assert` about a class is in fact the exception to the rule, as technically `TailShared!C` _should_ unshare the reference, but it doesn't, because it can't.

Reply:
> Is there a reason you are using the : operator instead of == inside the is expression?

`TailShared!S3` is not `shared S3` but a wrapper that converts implicitly. And that's what `:` means. The point of the wrapper is that the (property-wrapped) fields get tail-shared types. You get `i` as an `int` instead of a `shared(int)`, and `p` as a `shared(int)*` instead of a `shared(int*)`

> So the == in the assert about a class is in fact the exception to the rule, as technically TailShared!C should unshare the reference, but it doesn't, because it can't.

It's true that `TailShared!C` would strip shared from the reference if it could, but we'd test that with `==`.

`:` is the exception. It's only needed when `TailShared!S` resolves to the wrapper. And the wrapper is only needed for a certain kind of structs, not all of them.

Reply:
> `:` is the exception. It's only needed when `TailShared!S` resolves to the wrapper. And the wrapper is only needed for a certain kind of structs, not all of them.

By '`==` is the exception' I just meant that `is(TailShared!T == shared T)` if and only if T is a reference type (class or interface), while `is(TailShared!T : shared T)` for any type T.